### PR TITLE
Fix mempool cmd

### PR DESCRIPTION
**Pull request by abi87**
p2p command "mempool" was not duly update upon certificate introduction and was not listing certificates hashes to nodes asking for them.
This PR fixes the issue and, by means of some refactoring to CNode monster class, unit tests the "mempool" cmd processing

Note: initial target branch set to proof_sys_integration_base to ease up the review

### CODE REVIEW

File: src/net.cpp [hunk 1]
```diff
@@ -2089,12 +2089,8 @@ void Relay(const CTransactionBase& tx, const CDataStream& ss)
...
         if(!pnode->fRelayTxes)
             continue;
-        LOCK(pnode->cs_filter);
-        if (pnode->pfilter)
-        {
-            if (pnode->pfilter->IsRelevantAndUpdate(tx))
-                pnode->PushInventory(inv);
-        } else
+
+        if (pnode->nodeFilter.updateWith(tx))
```
Comment (i-Alex) [inline src/net.cpp:2093]: Seems a bit tricky that `updateWith(...)` returns `true` if `pfilter == nullptr`.
I would suggest to return `false` similar to the `insert(...)` method logic.
So, here we may do first: `if(!pnode->nodeFilter.isNull())) ...` and keep exactly the same flow as before.

**Review by i-Alex — CHANGES_REQUESTED** [review]

**Review by alsala — CHANGES_REQUESTED** [review]

File: src/net.cpp (hunk 1 above)
Comment (abi87) [reply src/net.cpp:2093]: Done

File: src/bloom.cpp [hunk 2]
```diff
@@ -271,3 +271,55 @@ void CRollingBloomFilter::reset()
     b2.reset(nNewTweak);
     nInsertions = 0;
 }
+
+CNodeFilter::CNodeFilter(): pfilter(nullptr)
+{
+    pfilter = new CBloomFilter();
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/bloom.cpp:277]: Done, deleting copy ctor and assignment operator

File: src/net.h [hunk 3]
```diff
@@ -263,6 +263,10 @@ class CNodeInterface //Interface useful for UTs
     virtual void StopAskingFor(const CInv& inv) = 0;
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/net.h:263]: Done

**Comment by ptagl** [comment]
These changes have already been integrated with [PR 122](https://github.com/HorizenOfficial/zend_oo/pull/122).

**Pull request closed without merging**